Objects/listobject.c

```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
         lo.values = saved_ob_item;
     }
 
+
+    /* The pre-sort check: here's where we decide which compare function to use.
+     * How much optimization is safe? We test for homogeneity with respect to 
+     * several properties that are expensive to check at compare-time, and 
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
+        /* Assume the first element is representative of the whole list. */
+        int keys_are_in_tuples = (lo.keys[0]->ob_type == &PyTuple_Type &&
```

Comment:
I could do that. I just wanted it to be consistent with the rest of the code, which also uses `ob_type` compares. But I think that would probably be closer to the current implementation; at least, `PyTuple_RichCompare` (or whatever it's called) does that (IIRC). So I can make this change.